### PR TITLE
Fixed #9061 -- Allowed FormSets to disable deleting extra forms.

### PR DESCRIPTION
**Pull request by smithdc1**
[Ticket #9061](https://code.djangoproject.com/ticket/9061)

This carries on work made in #4772 and #12558. I've updated the PR for current master and added the feature to model and inline formset factories as well. I've tried to layout the commits in a logical order - the first is intended to be the PR as per the previous attempt in 12558 with my amends in the subsequent commits. I hope this is useful. 

I'll leave a few comments inline shortly.

### CODE REVIEW

File: docs/topics/forms/formsets.txt [hunk 1]
```diff
@@ -22,7 +22,7 @@ a formset out of an ``ArticleForm`` you would do::
     >>> ArticleFormSet = formset_factory(ArticleForm)
 
 You now have created a formset class named ``ArticleFormSet``.
-Instantiating the formset gives you the ability to iterate over the forms 
+Instantiating the formset gives you the ability to iterate over the forms
```
Comment (smithdc1) [inline docs/topics/forms/formsets.txt:25]: I'm so sorry. I've spent a fair amount of time trying to revert this space change and have failed. I don't seem to be able to get git to pick up adding back the trailing space. 😢

File: tests/forms_tests/tests/test_formsets.py [hunk 2]
```diff
@@ -1197,6 +1197,61 @@ def test_html_safe(self):
         self.assertTrue(hasattr(formset, '__html__'))
         self.assertEqual(str(formset), formset.__html__())
 
+    def test_can_delete_extra_formset_forms(self):
```
Comment (smithdc1) [inline tests/forms_tests/tests/test_formsets.py:1200]: I didn't know if this test is required - there are plenty of other tests for deleting `formsets`. However, part of me liked the symmetry of having this test next to the following one to show the different behavior.

File: tests/forms_tests/tests/test_formsets.py [hunk 3]
```diff
@@ -1197,6 +1197,61 @@ def test_html_safe(self):
...
+        GenericFormFormset = formset_factory(
+            form=GenericForm,
+            can_delete=True,
+            extra=2)
+        formset = GenericFormFormset()
+        self.assertEqual(len(formset), 2)
+        self.assertIn('DELETE', formset.forms[0].fields)
+        self.assertIn('DELETE', formset.forms[1].fields)
+
+    def test_delete_fields_are_not_added_for_extra_formset_forms(self):
```
Comment (smithdc1) [inline tests/forms_tests/tests/test_formsets.py:1213]: I left as the original name, but could be `test_can_not_delete_extra_formset_forms` ?

**Comment by smithdc1** [comment]
Unclear why tests are failing. e.g. docs `ERROR: Error fetching remote repo 'origin'`

File: tests/forms_tests/tests/test_formsets.py (hunk 2 above)
Comment (carltongibson) [reply tests/forms_tests/tests/test_formsets.py:1200]: Let's keep it.

File: tests/forms_tests/tests/test_formsets.py (hunk 3 above)
Comment (carltongibson) [reply tests/forms_tests/tests/test_formsets.py]: I renamed the test and removed the docstring.

File: docs/topics/forms/formsets.txt (hunk 1 above)
Comment (carltongibson) [reply docs/topics/forms/formsets.txt:25]: Let's just leave it.

File: docs/ref/forms/models.txt [hunk 4]
```diff
@@ -70,12 +70,12 @@ Model Form API reference. For introductory material about model forms, see the
 
     .. versionchanged:: 3.2
 
-        The ``absolute_max`` argument was added.
+        The ``absolute_max`` and ``can_delete_extra`` arguments were added.
```
Comment (felixxm) [inline docs/ref/forms/models.txt:73]: We should add `can_delete_extra` to [the above list of arguments](https://github.com/django/django/pull/13097/files#diff-ad82f031e7988d81829bb4cf25a58dbbR64-R67), e.g.
```diff
diff --git a/docs/ref/forms/models.txt b/docs/ref/forms/models.txt
index e18660b3c2..aec3d977a3 100644
--- a/docs/ref/forms/models.txt
+++ b/docs/ref/forms/models.txt
@@ -62,8 +62,9 @@ Model Form API reference. For introductory material about model forms, see the
     through to :func:`~django.forms.models.modelform_factory`.
 
     Arguments ``formset``, ``extra``, ``max_num``, ``can_order``,
-    ``can_delete``, ``validate_max``, and ``absolute_max`` are passed through
-    to :func:`~django.forms.formsets.formset_factory`. See :doc:`formsets
+    ``can_delete``, ``can_delete_extra``, ``validate_max``, and
+    ``absolute_max`` are passed through to
+    :func:`~django.forms.formsets.formset_factory`. See :doc:`formsets
     </topics/forms/formsets>` for details.
 
     See :ref:`model-formsets` for example usage.
```

File: tests/forms_tests/tests/test_formsets.py [hunk 5]
```diff
@@ -1197,6 +1197,57 @@ def test_html_safe(self):
...
         self.assertEqual(str(formset), formset.__html__())
 
+    def test_can_delete_extra_formset_forms(self):
+        class GenericForm(Form):
+            foo = CharField()
+
+        GenericFormFormset = formset_factory(
+            form=GenericForm,
+            can_delete=True,
+            extra=2)
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:1207]: Please use hanging indentation:
```python
        GenericFormFormset = formset_factory(
            form=GenericForm,
            can_delete=True,
            extra=2,
         )
```

File: tests/forms_tests/tests/test_formsets.py [hunk 6]
```diff
@@ -1197,6 +1197,57 @@ def test_html_safe(self):
         self.assertTrue(hasattr(formset, '__html__'))
         self.assertEqual(str(formset), formset.__html__())
 
+    def test_can_delete_extra_formset_forms(self):
+        class GenericForm(Form):
+            foo = CharField()
+
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:1203]: Do we need a new form? I would reuse `Choice` in both tests:
```python
ChoiceFormFormset = formset_factory(Choice, can_delete=True, extra=2)
```

File: tests/forms_tests/tests/test_formsets.py [hunk 7]
```diff
@@ -1197,6 +1197,57 @@ def test_html_safe(self):
...
+    def test_disable_delete_extra_formset_forms(self):
+        class GenericForm(Form):
+            foo = CharField()
+
+        GenericFormFormset = formset_factory(
+            form=GenericForm,
+            can_delete=True,
+            can_delete_extra=False,
+            extra=2)
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:1221]: Hanging indentation.

File: tests/model_formsets/tests.py [hunk 8]
```diff
@@ -1916,3 +1916,57 @@ def test_inlineformset_factory_absolute_max_with_max_num(self):
...
             ['Please submit 20 or fewer forms.'],
         )
+
+    def test_modelformset_factory_can_delete_extra(self):
+        AuthorFormSet = modelformset_factory(
+            Author,
+            fields='__all__',
+            can_delete=True,
+            can_delete_extra=True,
+            extra=2
```
Comment (felixxm) [inline tests/model_formsets/tests.py:1926]: Add trailing comma (in all tests).

File: tests/forms_tests/tests/test_formsets.py [hunk 9]
```diff
@@ -1197,6 +1197,57 @@ def test_html_safe(self):
...
+            'form-TOTAL_FORMS': '3',
+            'form-INITIAL_FORMS': '1',
+        }, initial=[{'foo': 'bar'}])
+
+        self.assertEqual(formset.cleaned_data, [
+            {'foo': 'bar', 'DELETE': True},
+            {'foo': 'baz'},
+            {},
+        ])
+        self.assertTrue(formset._should_delete_form(formset.forms[0]))
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:1247]: `self.assertIs(..., True/False)`

**Comment by smithdc1** [comment]
Hi Both - I've added a new commit to this for the suggested changes. Hope I've captured them correctly.

Thanks all.

**Comment by smithdc1** [comment]
@felixxm your attention to detail is outstanding. :+1 thank you.

**Pull request merged**